### PR TITLE
Distance to nearest utility

### PR DESCRIPTION
**Pull request by GitCodeSM**
- changed nearest.py for the get the nearest bus stops function
- changed UtilityLocation.BUS_STOP string "bus stop" to "bus depot" for accuracy
- removed launch.json

### CODE REVIEW

File: locesh/distanceto/main.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+'''
+User,api authentication generates token_type,access_token
+'''
+
+import requests
+
+def mapmyindia_access_tokens(client_id: str, client_secret: str) -> tuple[str, str]:
```
Comment (sutyum) [inline locesh/distanceto/main.py:7]: `Tuple` not `tuple`. It needs to be imported from typing

File: locesh/distanceto/nearest.py [hunk 2]
```diff
@@ -1,33 +1,81 @@
...
+    token_type, access_token, lat, lon, keywords
+) -> List[UtilDistance]:
     """
-    Find the 5 nearest bus stops to a given location.
+    List of nearby places for a category
     """
-    return []
+
+    # location string to be concatenated
+    loc_str = str(lat) + "," + str(lon)
```
Comment (sutyum) [inline locesh/distanceto/nearest.py:15]: use type annotations

File: locesh/distanceto/nearest.py [hunk 3]
```diff
@@ -1,33 +1,81 @@
...
+        + loc_str
+        + "&itemCount=5"
+    )
+
+    token_dict: Dict = {
+        "token_type": token_type,
+        "access_token": access_token,
+    }
+
+    places_json_obj = requests.get(api_loc_url, token_dict).json()
```
Comment (sutyum) [inline locesh/distanceto/nearest.py:30]: use pydantic, use type annotation

File: locesh/distanceto/nearest.py [hunk 4]
```diff
@@ -1,33 +1,81 @@
...
+
+    # 5 nearby places being mapped to class UtilDistance
+    try:
```
Comment (sutyum) [inline locesh/distanceto/nearest.py:33]: completely unnecessary, use pydantic

File: locesh/distanceto/nearest.py [hunk 5]
```diff
@@ -1,33 +1,81 @@
...
+        items = [
+            UtilDistance(
+                **{
+                    "place_distance": x["distance"],
```
Comment (sutyum) [inline locesh/distanceto/nearest.py:37]: you should never do a keyword lookup on API respones. Use pydantic correctly

File: locesh/distanceto/nearest.py [hunk 6]
```diff
@@ -1,33 +1,81 @@
...
+                    "keywords": x["keywords"],
+                    "order_index": x["orderIndex"],
+                    "place_address": x["placeAddress"],
+                    "place_name": x["placeName"],
+                    "place_type": x["type"],
+                }
+            )
+            for x in places_json_obj["suggestedLocations"][0:5]
+        ]
+    except:
```
Comment (sutyum) [inline locesh/distanceto/nearest.py:48]: except what? `except:` as well as `except Exception:` are poor coding practice. Always be explicit about the exceptions you are trying to catch. In this case it should have been `except KeyError as e:`

File: locesh/distanceto/nearest.py [hunk 7]
```diff
@@ -1,33 +1,81 @@
...
+        print("Exception: Key error. key 'suggestedLocations' not found.")
```
Comment (sutyum) [inline locesh/distanceto/nearest.py:49]: catching an exception just to print something is almost never useful

File: locesh/distanceto/types.py [hunk 8]
```diff
@@ -1,16 +1,23 @@
-from dataclasses import dataclass
+'''
```
Comment (sutyum) [inline locesh/distanceto/types.py:1]: docstring not useful or informative

File: locesh/distanceto/types.py [hunk 9]
```diff
@@ -1,16 +1,23 @@
...
+#     lattitude: float
 
 
 class UtilityLocation(str, Enum):
-    BUS_STOP = "bus_stop"
+    '''
+    To provide search keyword as utility
+    '''
+
+    BUS_STOP = "bus stop"
```
Comment (sutyum) [inline locesh/distanceto/types.py:20]: ideally enum string should not contain a space character. prefer "bus_stop"

File: locesh/distanceto/types.py [hunk 10]
```diff
@@ -22,9 +29,23 @@ class UtilityLocation(str, Enum):
...
+
+
+class UtilDistance(BaseModel):
+    '''
+    For details list of nearest keyword places
+    '''
+
+    place_distance: str
+    place_loc: str
+    keywords: list[str] = []
```
Comment (sutyum) [inline locesh/distanceto/types.py:47]: `keywords: list[str] = []` bad practice, use `keywords: list[str] = Field(default_factory=list)` instead

File: locesh/distanceto/types.py [hunk 11]
```diff
@@ -22,9 +29,23 @@ class UtilityLocation(str, Enum):
...
+    order_index: str
```
Comment (sutyum) [inline locesh/distanceto/types.py:48]: why is this a string?

File: locesh/distanceto/types.py [hunk 12]
```diff
@@ -22,9 +29,23 @@ class UtilityLocation(str, Enum):
...
-    distance: float
-    location: Location
-    category: UtilityLocation
+# @dataclass
+# class UtilDistance:
+#     name: str
+#     distance: float
+#     location: Location
+#     category: UtilityLocation
+
```
Comment (sutyum) [inline locesh/distanceto/types.py:38]: missing pydantic model to parse the entire response with list of n UtilDistances

File: locesh/server/api.py [hunk 13]
```diff
@@ -1,25 +1,46 @@
...
 from distanceto.types import UtilDistance, UtilityLocation
 from server.app import app
 from server.auth_api import api_key_auth
 from server.types import MessageResponse
 
 
+@app.on_event("startup")
+async def startup_event(
+    client_id: str = client_id, client_secret: str = client_secret
+) -> tuple[str, str]:
```
Comment (sutyum) [inline locesh/server/api.py:14]: event triggers should return `None`

File: locesh/server/api.py [hunk 14]
```diff
@@ -1,25 +1,46 @@
...
 from distanceto.nearest import fetch_nearest_by_type
 from distanceto.types import UtilDistance, UtilityLocation
 from server.app import app
 from server.auth_api import api_key_auth
 from server.types import MessageResponse
 
 
+@app.on_event("startup")
+async def startup_event(
+    client_id: str = client_id, client_secret: str = client_secret
```
Comment (sutyum) [inline locesh/server/api.py:13]: event triggers should not require any inputs. These two values should come from the env variables

File: locesh/server/api.py [hunk 15]
```diff
@@ -1,25 +1,46 @@
...
+) -> tuple[str, str]:
+    # TODO: Use function from main module
+    token_type, access_token = mapmyindia_access_tokens(client_id, client_secret)
+    return token_type, access_token
```
Comment (sutyum) [inline locesh/server/api.py:17]: should not be returning anything

File: locesh/server/api.py [hunk 16]
```diff
@@ -1,25 +1,46 @@
...
 from server.auth_api import api_key_auth
 from server.types import MessageResponse
 
 
+@app.on_event("startup")
+async def startup_event(
+    client_id: str = client_id, client_secret: str = client_secret
+) -> tuple[str, str]:
+    # TODO: Use function from main module
+    token_type, access_token = mapmyindia_access_tokens(client_id, client_secret)
```
Comment (sutyum) [inline locesh/server/api.py:16]: you should have a seperate file with these two variables (eg: map_my_india.py), and these values should be imported here from that file. 

```py
# map_my_india.py
TOKEN_TYPE: str = ""
ACCESS_TOKEN: str = ""
```

then set values to these variables in the startup event trigger. During any api calls to mapmyindia api, use these values by importing these variables from map_my_india.py file

File: locesh/server/api.py [hunk 17]
```diff
@@ -1,25 +1,46 @@
...
 @app.get(
     "/distanceto/{utility}",
     tags=["Metrics"],
     description="Distance to a category of locations such as bus stop, school, etc.",
     dependencies=[Depends(api_key_auth)],
 )
 async def distanceto(
-    utility: UtilityLocation = Path(None, description="Location category of interest"),
-    lon: float = Query(None, description="Longitude of the location of interest"),
+    token_type: str = token_type,
```
Comment (sutyum) [inline locesh/server/api.py:30]: not needed, also not correct

File: locesh/server/api.py [hunk 18]
```diff
@@ -1,25 +1,46 @@
...
+@app.on_event("startup")
+async def startup_event(
+    client_id: str = client_id, client_secret: str = client_secret
+) -> tuple[str, str]:
+    # TODO: Use function from main module
+    token_type, access_token = mapmyindia_access_tokens(client_id, client_secret)
+    return token_type, access_token
+
+
+token_type, access_token = startup_event(client_id, client_secret)
```
Comment (sutyum) [inline locesh/server/api.py:20]: you are not supposed to call this function, fastapi calls it internally when the server starts

File: locesh/server/api.py [hunk 19]
```diff
@@ -1,25 +1,46 @@
...
     # TODO: Use function from distanceto module
-    distances = fetch_nearest_by_type(utility, lon=lon, lat=lat)
+    distances = fetch_nearest_by_type(
+        token_type=token_type,
+        access_token=access_token,
+        lat=lat,
+        lon=lon,
+        utility=utility,
+    )
     print(distances)
```
Comment (sutyum) [inline locesh/server/api.py:44]: comment out prints

**Review by sutyum — CHANGES_REQUESTED** [review]
Lots of issues, i have provided detailed comments

File: locesh/distanceto/nearest.py [hunk 20]
```diff
@@ -1,33 +1,81 @@
...
+
+    api_loc_url = (
+        "https://atlas.mapmyindia.com/api/places/nearby/json?explain&richData&&keywords="
+        + keywords
+        + "&refLocation="
+        + loc_str
+        + "&itemCount=5"
+    )
+
+    token_dict: Dict = {
```
[the remark this answers is not in the document]
Comment (GitCodeSM) [reply locesh/distanceto/nearest.py:25]: changed to 
token_dict: Dict[str, str]

File: locesh/distanceto/nearest.py [hunk 21]
```diff
@@ -1,33 +1,81 @@
...
+    List of nearby places for a category
     """
-    return []
+
+    # location string to be concatenated
+    loc_str = str(lat) + "," + str(lon)
+
+    api_loc_url = (
+        "https://atlas.mapmyindia.com/api/places/nearby/json?explain&richData&&keywords="
+        + keywords
```
[the remark this answers is not in the document]
Comment (GitCodeSM) [reply locesh/distanceto/nearest.py:19]: keywords changed to a list of strings transformed as a string
keywords: str = ",".join(utility)

File: locesh/distanceto/nearest.py [hunk 22]
```diff
@@ -1,33 +1,81 @@
 from distanceto.types import UtilDistance, UtilityLocation
+from typing import List, Dict
+import requests
+import json
 
 
-def get_closest_bus_stops(lon: float, lat: float) -> list[UtilDistance]:
+def mapmyindia_nearby_places(
+    token_type, access_token, lat, lon, keywords
```
[the remark this answers is not in the document]
Comment (GitCodeSM) [reply locesh/distanceto/nearest.py:8]: types provided in fetch_nearest... function

File: locesh/distanceto/nearest.py (hunk 6 above)
Comment (GitCodeSM) [reply locesh/distanceto/nearest.py:48]: changed to KeyError as e

File: locesh/distanceto/nearest.py (hunk 7 above)
Comment (GitCodeSM) [reply locesh/distanceto/nearest.py:49]: changed

File: locesh/distanceto/types.py (hunk 8 above)
Comment (GitCodeSM) [reply locesh/distanceto/types.py:1]: changed

File: locesh/distanceto/types.py (hunk 10 above)
Comment (GitCodeSM) [reply locesh/distanceto/types.py:47]: Can't change: Shows error:
Incompatible types in assignment (expression has type "Field[List[_T]]", variable has type "List[str]")  [assignment]mypy(error)

File: locesh/distanceto/types.py (hunk 11 above)
Comment (GitCodeSM) [reply locesh/distanceto/types.py:48]: all results of api are strings

File: locesh/server/api.py (hunk 13 above)
Comment (GitCodeSM) [reply locesh/server/api.py:14]: ok. changed to none

File: locesh/server/api.py (hunk 14 above)
Comment (GitCodeSM) [reply locesh/server/api.py:13]: changed the startup event

File: locesh/server/api.py (hunk 15 above)
Comment (GitCodeSM) [reply locesh/server/api.py:17]: ok. removed returns

File: locesh/server/api.py (hunk 16 above)
Comment (GitCodeSM) [reply locesh/server/api.py:16]: ok. changed. Now tokens are imported from main.py that has mapmyindia... token returning function

File: locesh/server/api.py (hunk 18 above)
Comment (GitCodeSM) [reply locesh/server/api.py:20]: ok. removed

File: locesh/server/api.py (hunk 19 above)
Comment (GitCodeSM) [reply locesh/server/api.py]: commented the print

File: locesh/distanceto/main.py (hunk 1 above)
Comment (GitCodeSM) [reply locesh/distanceto/main.py:7]: changed

File: locesh/distanceto/nearest.py [hunk 23]
```diff
@@ -1,33 +1,81 @@
...
+            for x in places_json_obj["suggestedLocations"][0:5]
+        ]
+    except:
+        print("Exception: Key error. key 'suggestedLocations' not found.")
+
+    return items
 
 
-def get_closest_schools(lon: float, lat: float) -> list[UtilDistance]:
+def get_closest_bus_stops(
```
[the remark this answers is not in the document]
Comment (GitCodeSM) [reply locesh/distanceto/nearest.py:54]: commented out and changed in fetch_nearest... function as well

File: locesh/distanceto/types.py (hunk 9 above)
Comment (GitCodeSM) [reply locesh/distanceto/types.py:20]: ok

File: locesh/server/api.py (hunk 16 above)
Comment (GitCodeSM) [reply locesh/server/api.py]: When I assign imported values to variables in this trigger event, it gives an error as variables aren't used anywhere. Also, I removed inputs and set the return type to none for the start event as you said in one of the reviews.

File: locesh/distanceto/main.py (hunk 1 above)
Comment (GitCodeSM) [reply locesh/distanceto/main.py:10]: but gave error:
distanceto\main.py:10:69: UP006 Use `tuple` instead of `Tuple` for type annotations

Comment (GitCodeSM) [reply locesh/distanceto/main.py:10]: fixing error reverted it back to tuple

File: locesh/distanceto/nearest.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (GitCodeSM) [reply locesh/distanceto/nearest.py]: ok

File: locesh/server/api.py (hunk 17 above)
Comment (GitCodeSM) [reply locesh/server/api.py]: removed

**Pull request closed without merging**